sled-agent/src/instance.rs

```diff
@@ -526,25 +530,55 @@ impl InstanceInner {
     /// This routine is safe to call even if the instance's zone was never
     /// started. It is also safe to call multiple times on a single instance.
     async fn terminate(&mut self) -> Result<(), Error> {
+        let zname = propolis_zone_name(self.propolis_id());
+
+        // First fetch the running state.
+        //
+        // If there is nothing here, then there is no `RunningZone`, and so
+        // there's no zone or resources to clean up at all.
+        let mut running_state = if let Some(state) = self.running_state.take() {
+            state
+        } else {
+            debug!(
+                self.log,
+                "Instance::terminate() called with no running state"
+            );
+            return Ok(());
+        };
+
+        // Take a zone bundle whenever this instance stops.
+        let context = self
+            .storage
+            .zone_bundle_context(&zname, ZoneBundleCause::TerminatedInstance)
+            .await;
+        if let Err(e) = zone_bundle::create(
+            &self.log,
+            &running_state.running_zone,
+            &context,
+        )
+        .await
+        {
+            error!(
+                self.log,
+                "Failed to take zone bundle for terminated instance";
+                "zone_name" => &zname,
+                "reason" => ?e,
+            );
```

Comment:
So, on the one hand, doing this is great, because we'll capture a lot more info. On the other hand, it seems like it'll massively inflate the amount of traffic moving onto the M.2s (and U.2s, if I understand #3713 correctly).

Do we ever trim down the set of zone bundles? E.g., removing the oldest ones when we fill up too much space?

(honestly all other comments on this PR are nits, this is the one that IMO matters most)

Reply:
Other than calls to `zone-bundle delete`, we do not ever remove bundles. That's an in-progress PR I'll be putting up next. I agree it's critical -- I debated putting it in this PR, but it was already borderline in terms of size for a review.

Reply:
Also, an argument could be made for _not_ taking automatic instance zone bundles here until the cleanup stuff is in place. I think enough folks have lost the race against the zone-cleanup machinery when an instance dies that it would be valuable. But it's certainly not a clear-cut decision.

Reply:
Yup, we can land them separately, but I wouldn't want to ship this PR to customers without that follow-up, since that'll make us hit our resource limits with no reprieve. 

However, if that's already on your radar, I'm okay punting it -- as long as we're on the same page about urgency.